src/transformers/generation/utils.py

```diff
@@ -1922,11 +1925,24 @@ def _contrastive_search(
         logits_warper = logits_warper if logits_warper is not None else LogitsProcessorList()
         stopping_criteria = stopping_criteria if stopping_criteria is not None else StoppingCriteriaList()
         pad_token_id = pad_token_id if pad_token_id is not None else self.generation_config.pad_token_id
-        eos_token_id = eos_token_id if eos_token_id is not None else self.generation_config.eos_token_id
-        sequential = sequential if sequential is not None else self.generation_config.low_memory
+        if eos_token_id is not None:
+            warnings.warn(
+                "`eos_token_id` is deprecated in this function and will be removed in v4.41, use"
+                " `stopping_criteria=StoppingCriteriaList([EOSTokenCriteria(eos_token_id=eos_token_id)])` instead.",
+                FutureWarning,
+            )
+            stopping_criteria.append(EOSTokenCriteria(eos_token_id=eos_token_id))
+        else:
+            eos_token_id = [
+                criteria.eos_token_id for criteria in stopping_criteria if hasattr(criteria, "eos_token_id")
+            ]
+            if not eos_token_id and self.generation_config.eos_token_id:
+                stopping_criteria.append(EOSTokenCriteria(eos_token_id=eos_token_id))
+                eos_token_id = self.generation_config.eos_token_id
```

Comment:
```suggestion
        if eos_token_id is not None:
            warnings.warn(
                "`eos_token_id` is deprecated in this function and will be removed in v4.41, use"
                " `stopping_criteria=StoppingCriteriaList([EOSTokenCriteria(eos_token_id=eos_token_id)])` instead.",
                FutureWarning,
            )
            stopping_criteria.append(EOSTokenCriteria(eos_token_id=eos_token_id))
```
(see 1. my PR-wide comment)
2. If 1. is applied, we don't need to set `eos_token_id`. As such, we don't need the `else` case at all (`eos_token_id` is `None` in the `else`, so the associated stopping criteria shouldn't exist)

Reply:
Yes, we do not need if the methods are called from `generate`. But I thought of two cases that might lead to errors, because users are still able to call generate methods directly:

1. If they do not pass in "eos" explicitly, neither they pass the new `EOSTokenCriteria`, their generation might not stop at "eos" going until max length. Worse, if they do not pass any StoppingCriteria, then generation will go forever. Before this PR, we always took generation config's "eos" by default and used it to stop generation.
2.  Beam search methods need an "eos" for their scorer's correct functioning. We discussed in one of prev PRs and decided that calling the criterias twice in one loop is not good (but theoretically still possible), so we have to define `eos_token_id` in the beginning anyway. At least until beam search methods are refactored

So the question is, shouldn't we be more cautious until the methods become completely private? Or maybe leave it only for beam search methods, removing completely for other methods?

Reply:
You're right, it needs more thought than my code suggestion. Another thing we haven't considered is the case where the user passes both `eos_token_id` and the new stopping criteria.

Let's keep the `else` case with a TODO to remove it after beam search is refactored 👍 Let's also add a case in this if/else: if the new stopping criteria is already present, let's not add a duplicate.